overlord/snapstate/refreshhints.go

```diff
@@ -195,9 +197,26 @@ func refreshHintsFromCandidates(st *state.State, updates []*snap.Info, ignoreVal
 // pruneRefreshCandidates removes the given snaps from refresh-candidates map
 // in the state.
 func pruneRefreshCandidates(st *state.State, snaps ...string) error {
+	tr := config.NewTransaction(st)
+	gateAutoRefreshHook, err := features.Flag(tr, features.GateAutoRefreshHook)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	// do nothing if gate-auto-refresh-hook feature is not enabled.
+	// this acts as a workaround for the case where a snapd from edge was
+	// used and created refresh-candidates in the old format (an array) with the
+	// feature enabled, but the feature was then disabled so the new map format
+	// will never make it into the state. When the feature is enabled then
+	// auto-refresh code will re-initialize refresh-candidates in the correct format
+	// expected here.
+	// See https://forum.snapcraft.io/t/cannot-r-emove-snap-json-cannot-unmarshal-array-into-go-value-of-type-map-string-snapstate-refreshcandidate/27276
+	if !gateAutoRefreshHook {
```

Comment:
don't we have the issue that refresh-candidates might contain stale values when the feature is re-enabled if we don't prune it all the time?

Reply:
Yes, but they aren't going to be used anywhere since only the new auto-refresh code uses them.

Reply:
I changed this code to remove them from state.